### PR TITLE
Add support for loading custom formatter from package

### LINKED ISSUE #6100: Add support for PnP as linked customFormatters
### What steps are needed to reproduce the bug?

Run `yarn stylelint --custom-formatter mypackage/stylelint-formatter ...`

### What Stylelint configuration is needed to reproduce the bug?

```
{
  "extends": ["@equisoft/stylelint-config"]
}
```

### How did you run Stylelint?

CLI with `stylelint --custom-formatter mypackage/stylelint-formatter src/browser/**/*.{css,scss,js,jsx,ts,tsx}` 

### Which version of Stylelint are you using?

14.8.2

### What did you expect to happen?

The formatter is loaded from the dependency.

### What actually happened?

stylelint prepends `process.cwd()` to the provided custom-formatter option

### Does the bug relate to non-standard syntax?

No

### Proposal to fix the bug

_No response_

### PR DESCRIPTION
**Pull request by meriouma**
> Which issue, if any, is this issue related to?

Closes #6100

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: lib/resolveCustomFormatter.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+		return formatterPath;
+	}
+
+	const relativePath = path.join(process.cwd(), formatterPath);
+
+	if (fs.existsSync(relativePath)) {
+		return relativePath;
+	}
+
+	return formatterPath;
```
Comment (JounQin) [inline lib/resolveCustomFormatter.js:19]: It should call `require.resolve` here to ensure the path is correct.

**Review by JounQin — CHANGES_REQUESTED** [review]

**Review by JounQin — APPROVED** [review]

File: lib/resolveCustomFormatter.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * @param {string} formatterPath
+ * @returns {string}
+ */
+function resolveCustomFormatter(formatterPath) {
+	if (path.isAbsolute(formatterPath)) {
+		return formatterPath;
+	}
+
+	const relativePath = path.join(process.cwd(), formatterPath);
```
Comment (JounQin) [inline lib/resolveCustomFormatter.js:13]: I'd prefer to use `path.resolve` which covers the case for absolute path.

**Review by JounQin — CHANGES_REQUESTED** [review]

File: lib/resolveCustomFormatter.js (hunk 2 above)
Comment (JounQin) [inline lib/resolveCustomFormatter.js:13]:
```suggestion
	const relativePath = path.resolve(process.cwd(), formatterPath);
```

**Review by JounQin — CHANGES_REQUESTED** [review]

File: lib/resolveCustomFormatter.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+	if (fs.existsSync(relativePath)) {
+		return relativePath;
+	}
```
Comment (JounQin) [inline lib/resolveCustomFormatter.js:17]:
```suggestion
	if (fs.existsSync(formatterPath)) {
		return formatterPath;
	}
```

**Review by JounQin — CHANGES_REQUESTED** [review]

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+		const aModulePath = '@stylelint/prettier-config/index.js';
+
+		path.isAbsolute.mockReturnValue(false);
+		fs.existsSync.mockReturnValue(false);
+
+		const result = resolveCustomFormatter(aModulePath);
+
+		const realPathModule = jest.requireActual('path');
+		const expectedPath = realPathModule.join('node_modules', aModulePath);
+		expect(result).toEqual(expect.stringMatching(RegExp(`${expectedPath}$`)));
```
Comment (JounQin) [inline lib/__tests__/resolveCustomFormatter.test.js:32]:
```suggestion

		expect(result).toEqual(expect.stringMatching(RegExp(`${expectedPath}$`)));
```

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+		const result = resolveCustomFormatter(aRelativePath);
+
+		expect(result).toEqual(expected);
+	});
+
+	it('should return provided path when path is neither absolute nor relative', () => {
+		const aModulePath = '@stylelint/prettier-config/index.js';
+
+		path.isAbsolute.mockReturnValue(false);
```
Comment (JounQin) [inline lib/__tests__/resolveCustomFormatter.test.js:25]:
```suggestion
```

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+		fs.existsSync.mockReturnValue(false);
+
+		const result = resolveCustomFormatter(aModulePath);
+
+		const realPathModule = jest.requireActual('path');
+		const expectedPath = realPathModule.join('node_modules', aModulePath);
+
+		expect(result).toEqual(expect.stringMatching(RegExp(`${expectedPath}$`)));
```
Comment (JounQin) [inline lib/__tests__/resolveCustomFormatter.test.js:33]:
```suggestion
		expect(result).toEqual(expectedPath);
```

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+const fs = require('fs');
+const path = require('path');
+const { resolveCustomFormatter } = require('../resolveCustomFormatter');
+
+jest.mock('fs');
+jest.mock('path');
```
[the remark this answers is not in the document]
Comment (meriouma) [reply lib/__tests__/resolveCustomFormatter.test.js]: I believe this is inconsistent with the other test files.

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+	it('should return provided path when path is neither absolute nor relative', () => {
+		const aModulePath = '@stylelint/prettier-config/index.js';
+
+		fs.existsSync.mockReturnValue(false);
+
+		const result = resolveCustomFormatter(aModulePath);
+
+		const realPathModule = jest.requireActual('path');
+		const expectedPath = realPathModule.join('node_modules', aModulePath);
```
Comment (JounQin) [inline lib/__tests__/resolveCustomFormatter.test.js:30]:
```suggestion
		const expectedPath = realPathModule.resolve('node_modules', aModulePath);
```

**Comment by JounQin** [comment]
@meriouma Don't do suqash locally, keep the commits history, that's fine. We'll merge with squash.

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+	it('should return provided path when path is neither absolute nor relative', () => {
+		const aModulePath = '@stylelint/prettier-config/index.js';
+
+		fs.existsSync.mockReturnValue(false);
+
+		const result = resolveCustomFormatter(aModulePath);
+
+		const realPathModule = jest.requireActual('path');
+		const expectedPath = realPathModule.resolve('node_modules', aModulePath);
```
Comment (JounQin) [inline lib/__tests__/resolveCustomFormatter.test.js:30]:
```suggestion
		const expectedPath = realPathModule.resolve('lib/__tests__/node_modules', aModulePath);
```

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+		const aModulePath = '@stylelint/prettier-config/index.js';
+
+		fs.existsSync.mockReturnValue(false);
+
+		const result = resolveCustomFormatter(aModulePath);
+
+		const realPathModule = jest.requireActual('path');
+		const expectedPath = realPathModule.resolve('lib/__tests__/node_modules', aModulePath);
+
+		expect(result).toEqual(expectedPath);
```
Comment (JounQin) [inline lib/__tests__/resolveCustomFormatter.test.js:32]:
```suggestion
		const expectedPath = realPathModule.join('node_modules', aModulePath);

		expect(result.endsWith(expectedPath)).toBe(true);
```

**Review by JounQin — APPROVED** [review]

File: lib/cli.js [hunk 11]
```diff
@@ -338,9 +339,7 @@ module.exports = async (argv) => {
 	let formatter = cli.flags.formatter;
 
 	if (cli.flags.customFormatter) {
-		const customFormatter = path.isAbsolute(cli.flags.customFormatter)
-			? cli.flags.customFormatter
-			: path.join(process.cwd(), cli.flags.customFormatter);
+		const customFormatter = resolveCustomFormatter(cli.flags.customFormatter);
```
Comment (ybiquitous) [inline lib/cli.js:343]: Could you add a test case to cover the code calling `resolveCustomFormatter()` to `cli.test.js`?

**Review by ybiquitous — COMMENTED** [review]
The change looks good to me. I've left a suggestion for testing.

File: lib/__tests__/resolveCustomFormatter.test.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
+jest.mock('fs');
+jest.mock('path');
+
+const fs = require('fs');
+const path = require('path');
+const { resolveCustomFormatter } = require('../resolveCustomFormatter');
```
Comment (ybiquitous) [inline lib/__tests__/resolveCustomFormatter.test.js:6]:
```suggestion
const resolveCustomFormatter = require('../resolveCustomFormatter');
```

File: lib/cli.js [hunk 13]
```diff
@@ -13,6 +13,7 @@ const printConfig = require('./printConfig');
 const resolveFrom = require('resolve-from');
 const standalone = require('./standalone');
 const writeOutputFile = require('./writeOutputFile');
+const { resolveCustomFormatter } = require('./resolveCustomFormatter');
```
Comment (ybiquitous) [inline lib/cli.js:16]:
```suggestion
const resolveCustomFormatter = require('./resolveCustomFormatter');
```

File: lib/resolveCustomFormatter.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+	if (fs.existsSync(resolvedPath)) {
+		return resolvedPath;
+	}
+
+	return require.resolve(formatterPath);
+}
+
+module.exports = {
+	resolveCustomFormatter,
+};
```
Comment (ybiquitous) [inline lib/resolveCustomFormatter.js:20]:
```suggestion
module.exports = function resolveCustomFormatter(formatterPath) {
	const resolvedPath = path.resolve(formatterPath);

	if (fs.existsSync(resolvedPath)) {
		return resolvedPath;
	}

	return require.resolve(formatterPath);
};
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]
[suggest] Sorry, I missed one point. Could you make single exporting consistent with other modules?

**Review by ybiquitous — CHANGES_REQUESTED** [review]
[suggest] I think it's good to improve the CLI help and the doc about the `--custom-formatter` option. It will support a package name, not only a relative/absolute file path.

Could you update also the following files?

https://github.com/stylelint/stylelint/blob/743030d1bda238cfdb4b1aede456f34504d8824d/lib/cli.js#L178-L180

https://github.com/stylelint/stylelint/blob/743030d1bda238cfdb4b1aede456f34504d8824d/docs/user-guide/usage/options.md?plain=1#L91

**Comment by JounQin** [comment]
@meriouma Are you still working on this?

**Comment by meriouma** [comment]
@JounQin Yes, I will apply the changes. I stopped pushing changes yesterday so we were not two persons pushing in the branch at the same time.

**Comment by JounQin** [comment]
> @JounQin Yes, I will apply the changes. I stopped pushing changes yesterday so we were not two persons pushing in the branch at the same time.

You'll only need to pull the latest codes and then continue the work. 🍺

File: lib/__tests__/__snapshots__/cli.test.js.snap [hunk 15]
```diff
@@ -94,6 +94,7 @@ exports[`CLI --help 1`] = `
     --custom-formatter
 
       Path to a JS file exporting a custom formatting function.
+      The file can either be a filesystem path, or a file to load from a dependency.
```
Comment (JounQin) [inline lib/__tests__/__snapshots__/cli.test.js.snap:97]: A module name is also file.

**Review by JounQin — CHANGES_REQUESTED** [review]

**Comment by ybiquitous** [comment]
@meriouma Thank you for addressing the requested changes! The PR looks almost good to me.

I got an idea to test a custom formatter dependency in the repository: to install a private internal dependency. For example:

`package.json`:

```json
{
  "devDependencies": {
    "stylelint-test-custom-formatter": "file:packages/stylelint-test-custom-formatter"
  }
}
```

`packages/`:

```console
$ tree packages
packages/
└── stylelint-test-custom-formatter/
    ├── index.js
    └── package.json

1 directory, 2 files
```

Run a test:

```console
$ bin/stylelint.js --config=scripts/visual-config.json --custom-formatter=stylelint-test-custom-formatter scripts/*.css
Custom formatter reports 1 warning(s).%
```

What do you think about this idea?

**Comment by ybiquitous** [comment]
@JounQin I thought about how to test the dependency behavior without mocking. But mocking dependency resolution in `cli.test.js` may be enough and straightforward.

**Comment by JounQin** [comment]
> @JounQin I thought about how to test the dependency behavior without mocking. But mocking dependency resolution in `cli.test.js` may be enough and straightforward.

Yeah, a mock module like https://github.com/stylelint/stylelint/blob/main/lib/__tests__/__mocks__/get-stdin.js is fine IMO.

(Besides, https://github.com/stylelint/stylelint/blob/main/lib/__tests__/__mocks__/get-stdin.js seems can be deleted safely in favor of #6182

**Review by ybiquitous — APPROVED** [review]
Great! LGTM 👍🏼

**Review by JounQin — APPROVED** [review]

**Comment by JounQin** [comment]
@meriouma Thanks for your contribution!

**Comment by ybiquitous** [comment]
Changelog entry added:

> - Added: support for loading custom formatter from package ([#6228](https://github.com/stylelint/stylelint/pull/6228)).

**Pull request merged**